### ISSUE TITLE
Point features dropped from small dataset

### ISSUE THREAD

**Issue by timiyay** [mapbox/tippecanoe#166]
I have a small PostGIS table of ~2200 point locations for schools, which I have exported to a GeoJSON `FeatureCollection` using `ogr2ogr`. I'm trying to use `tippecanoe` to convert this data into vector `.mbtiles`, which I can upload to Mapbox.

However, once I upload this `.mbtiles` file via the Studio web app, most of the features are actually missing. When I inspect the `.mbtiles` file using `sqlite3`, the `tiles` table only contains 23 rows, even though the ~2200 features cover an area of 237,629 km² (the Australian state of Victoria).

I've experimented with different drop rates and such, with no change in outcome. Oddly, the final completion percentage varies between 99.9% 100.1%.

I'm using v1.6.4 of `tippecanoe`.

```bash
$ tippecanoe -f -o schools.mbtiles -B 0 -Z 0 -z 22 schools_dump.json
using layer 0 name schools_dump
Sorting 2234 features
2234 features, 28040 bytes of geometry, 22320 bytes of metadata, 55468 bytes of string pool
  99.9%  22/0/0
```
Is this expected behaviour?

**Background:**
I'm searching for a tool that I can use as part of an automated processing pipeline, to push data from a PostGIS DB to vector `.mbtiles`, and up to my Mapbox account. I've looked at the `awesome-vector-tiles` repo, and it's list of vector tile generators, but I don't get a clear picture of how to use these tools to create *Mapbox-compatible* MBtiles files.

I've gravitated towards `tippecanoe` for this purpose, as it's the best combo of user-friendly and automatable. Nonetheless, if I'm using it for the wrong purpose, and I should look to a different tool, I'd be all ears.

**Comment by ericfischer** [comment]
@timiyay, sorry for the confusing progress indicator. Because the tile generation is multithreaded, the threads sometimes disagree on exactly how far the generation has progressed.

My best guess is that your data is not projected in WGS84, so a lot of coordinates are perhaps getting clipped away and not actually included in the tiles. It should be easy to check this by visually inspecting the `.json` file: are the coordinates all in the range -180…180,-90…90? You may need to use `-t_srs EPSG:4326` in your `ogr2ogr` command to do this reprojection.

If the coordinates are reasonable, I would expect

    tippecanoe -f -o schools.mbtiles -B 0 schools_dump.json

to include all the features at every zoom level. Going all the way down to zoom level 22 seems very excessive: nothing needs the kind of precision that you get at that level.

**Comment by timiyay** [comment]
Thanks for the reply @ericfischer. I can confirm I'm outputting the geometry in WGS84 (I've attached a sample below).

```json
{
  "type": "FeatureCollection",
  "crs": {
    "type": "name",
    "properties": {
      "name": "urn:ogc:def:crs:OGC:1.3:CRS84"
    }
  },
  "features": [
    {
      "type": "Feature",
      "properties": {
        "sector": "Government",
        "name": "Alberton Primary School",
        "type": "Primary"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          146.666601,
          -38.617713
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "sector": "Government",
        "name": "Allansford and District Primary School",
        "type": "Primary"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          142.590393,
          -38.386281
        ]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "sector": "Government",
        "name": "Avoca Primary School",
        "type": "Primary"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [
          143.475649,
          -37.084502
        ]
      }
    }
  ]
}
```

Yes, the max zoom of 22 is excessive and, once I can get the `.mbtiles` working as I need, I'll be trimming down to zoom 17ish.

I'm yet to get `.mbtiles` output from `tippecanoe` that matches the data content and size I get from a manual GeoJSON upload via the Mapbox Studio web app.

**Comment by ericfischer** [comment]
Even z17 seems excessive, but I'll take your word that the precision is useful.

Can you run `tippecanoe-decode` on the `.mbtiles` file to get an idea of what is being dropped? When I run

    tippecanoe -f -o schools.mbtiles -B0 schools.json
    tippecanoe-decode schools.mbtiles

on your excerpt above, I see the expected 3 points at each zoom level, in a total of 34 tiles across 14 zooms.

By the way, if tippecanoe doesn't do what you want but direct GeoJSON upload does, you *can* use `mapbox-upload` to upload your GeoJSON directly from a script. I would still like to understand what is going wrong with tippecanoe though.

**Comment by ericfischer** [comment]
Actually if I say -z22 my tileset puts all the points at `0,0`, so that's the problem. I'll figure out what is broken about asking for such a high maxzoom. Must be a numeric precision problem somewhere.

**Comment by ericfischer** [comment]
Should be fixed now with https://github.com/mapbox/tippecanoe/commit/e7f571c9c35f6f725e85bd26e2a65f60245c925f in https://github.com/mapbox/tippecanoe/pull/167. Thank you for the report! Please let let me know if it works for you now.

**Comment by timiyay** [comment]
@ericfischer Thanks for the speedy update, much appreciated!

I hadn't been making use of `tippecanoe-decode` for debugging, so that's going to be very handy.

It's possible that my issues are not coming from `tippecanoe`, and are a result user error / bugs for MapboxGL. I'm chasing up that possibility in a separate issue (https://github.com/mapbox/mapbox-gl-js/issues/2054), and will update this issue when I get everything resolved.

**Comment by timiyay** [comment]
@ericfischer
My problem _does_ seem to be a product of creating my own vector tiles with `tippecanoe`, and not with MapboxGL. As such, it looks like the change in e7f571c didn't solve my issues.

I have a CodePen that shows the issue: http://codepen.io/anon/pen/YwOxaa

and here's a couple of screenshots that show it as well:
**Expected**
![expected](https://cloud.githubusercontent.com/assets/1470250/12805322/891b2936-cb4e-11e5-820f-bb4e8ac7c206.png)

**Actual**
![actual](https://cloud.githubusercontent.com/assets/1470250/12805328/91950ad2-cb4e-11e5-975f-9e2e7976b378.png)

**Expected (zoomed in)**
![expected-closer](https://cloud.githubusercontent.com/assets/1470250/12805333/987646ae-cb4e-11e5-9887-c289a0e09cef.png)

**Actual (zoomed in)**
![actual-closer](https://cloud.githubusercontent.com/assets/1470250/12805334/9de3901a-cb4e-11e5-8c24-56650abd3455.png)

Is is possible that I'm causing the problem by incorrect use of tippecanoe's CLI params? I'm still trying to grasp the concepts behind the dot dropping, and the choice of zoom levels. In the first instance, with such a small dataset, my intent is to create the minimum tilesets I need to display ~2200 points at all zoom levels. Once I'm confident that works, I can expand my method to larger polygon and line datasets.

**Comment by ericfischer** [comment]
To be clear, can you tell me exactly what tippecanoe command you are using to create these tiles?

Looking at the tilejson of `timiyay-redronin.7usjzkmt` it looks like your zoom levels for this tileset run from 0 to 19, so it is not exactly the -z22 command line you pasted above, but I can't tell what options you are specifying beyond that.

To try to get as close as I can to what you are doing, I downloaded a set of locations of South Australia schools from https://data.sa.gov.au/data/dataset/south-australian-government-education-and-child-development-sites-location-data-as-at-january-2015/resource/2bdd9cf6-0f59-4e24-9a37-3e35337d15e5 and converted them into this GeoJSON file: https://gist.github.com/ericfischer/42989b09e3c58255d634

I then ran that through Tippecanoe:

```
$ tippecanoe -f -B0 -o schools.mbtiles schools.geojson
For layer 0, using name "schoolsgeojson"
Sorting 2276 features
2276 features, 42917 bytes of geometry, 73153 bytes of metadata, 153482 bytes of string pool
  100.4%  14/14500/9897
```

This uses the default parameters to produce z0 through z14, which should be plenty of precision for any reasonable display. I uploaded the vector mbtiles as `enf.2a7kgsqq`.

I edited your CodePen file to use my mapid and to center on Adelaide instead of on Melbourne: http://codepen.io/anon/pen/PZdXwQ. As far as I can tell it does what it is supposed to, displaying all of the school locations at any zoom level.

Can you please try again, and if it still doesn't work for you, share your data file, command line, and resulting mbtiles?

**Comment by timiyay** [comment]
Thanks for comments and the Codepen. This is good stuff, as it makes user error more likely. A better result all around.

Happily, I can get the expected results when following your method. My issues seem to center around:  

* use of line-delimited JSON, and the `-P` flag
* misunderstanding the use of the CLI switches for `[-z maxzoom] [-Z minzoom] [-B basezoom]`

If I use default settings with `-B0`, then I get expected results

My source GeoJSON of Victorian schools:
https://gist.github.com/timiyay/5c7147dfc29d0b96b85e

My `tippecanoe` command:
`tippecanoe -f -B0 -o schools.mbtiles schools.geojson`

My CodePen, now functioning as expected:
http://codepen.io/anon/pen/QyVPeW

##### Line-delimited JSON

I've been experimenting with line-delimited JSON and the `-P` flag, to speed up processing for much-larger polygon datasets.

I convert my geojson to jsonlines with:
`jq -c '.features[]' schools.geojson > schools.jsonlines`

...which results in the following JSON: https://gist.github.com/timiyay/d1ee8efe0437d35aa389

Then, I compile my school tiles with:
`tippecanoe -f -P -o schools.mbtiles schools.jsonlines`

..which results in dropped points: http://codepen.io/anon/pen/GoYgvx

I also get the dropped points result if I run the command without the `-P` flag, using a jsonlines file as the source.

##### Z flags

I've also been getting dropped points when using `-z` and `-Z` flags. This is probably due to my misunderstanding of how the point dropping works. I was trying to use the min and max zoom on a larger polygon dataset to restrict the size and processing time of the dataset, and I mistakenly assumed it'd have little bearing on such a small point dataset.

If I render tiles from my geojson source using:
`tippecanoe -f -o schools.mbtiles -Z 0 -z 22 schools.geojson`

.. then I get the same point-dropping result.

Notably, when my points get dropped, it seems to happen in a repeatable fashion. When viewing in MapboxGL, the same points render in the same positions, implying that the point dropping is uniform.

Here is a Codepen with the tileset created with these flags: http://codepen.io/anon/pen/vLVEWp

-----------

@ericfischer  Thanks very much for showing me the simpler way to do things. I'll need to research what `-z` and `-Z` are actually doing, but I'll steer clear of them for now. Hopefully it all turns out to be user error.

**Comment by ericfischer** [comment]
Sorry for the lack of clarity around `-z` and `-Z`.

`-Z` straightforwardly sets the lowest zoom level that will be produced in the mbtiles. Normally it is 0 but you can set it if you don't want the low zooms for some reason.

`-z` is more complicated because, unless you also use `-B`, it sets both the highest zoom level that appears in the output *and* the "base" zoom level (the one at which all points are present). By default the base zoom is 14, so all features will be present at z14, 1/2.5 of them will be present at z13, 1/2.5^2 of them will be present at z12, etc. You are losing lots of features when you say `-z22` without `-B`, because you are saying that you want all features to be present at z22, 1/2.5 of them at z21, 1/2.5^2 at z20, etc, so there is almost nothing left by the time you get down to a zoom level that shows any significant area.

Note that you do not need to produce tiles down to z22 to be able to look at the map at z22. At any zoom beyond the maxzoom, it will use the data from the maxzoom tile. You only need to tile to a very high zoom level if you have extremely dense point data or need extremely detailed locations. Because your data is sparse and not extremely precise, you could get away with `-z12` or possibly even lower to make your processing fast and your tileset small.

If you want all points to appear at all zooms, you can either use `-B0` to set the base zoom level to 0, so all points are preserved at all zooms above 0, or `-r1` to set the dot-dropping rate to 1, so that all features are preserved at all zooms below the base zoom.

You may find it useful to use `-Bg -rg` to have Tippecanoe guess a base zoom and drop rate for you. In this case it will choose a base zoom of 0 because it considers 2200 points reasonable at any zoom level.

---

I hope you are getting exactly the same results from line-delimited JSON either with or without `-P`. If you aren't, that a bug. But it sounds like your missing features are from dot-dropping at zooms below the basezoom, not from a bug in parallel input.

**Comment by timiyay** [comment]
@ericfischer Once again, thanks for your guidance on this.

I'm happy to put this down as user error on my part, and close it off.

I'll keep an eagle-eye out for any unexpected behaviours, based on my new understanding of the CLIflags.
